File: src/arduinobot_remote/scripts/task_server.py
#!/usr/bin/env python3
import rospy
import actionlib
from arduinobot_remote.msg import ArduinobotTaskAction, ArduinobotTaskResult
import sys
import moveit_commander


class TaskServer(object):
    result = ArduinobotTaskResult()
    arm_goal = []
    gripper_goal = []

    def __init__(self, name):
        self.action_name_ = name
        moveit_commander.roscpp_initialize(sys.argv)
        self.arm_move_group_ = moveit_commander.MoveGroupCommander("arduinobot_arm")
        self.gripper_move_group_ = moveit_commander.MoveGroupCommander("arduinobot_hand")
        self.as_ = actionlib.SimpleActionServer(self.action_name_, ArduinobotTaskAction, execute_cb=self.execute_cb, auto_start=False)
        self.as_.start()

    def execute_cb(self, goal):
        success = True

        if goal.task_number == 0:
            self.arm_goal = [0.0, 0.0, 0.0]
            self.gripper_goal = [-0.7, -0.7]
        elif goal.task_number == 1:
            self.arm_goal = [-1.14, -0.6, -0.07]
            self.gripper_goal = [0.0, 0.0]
        elif goal.task_number == 2:
            self.arm_goal = [-1.57, 0.0, -1.0]
            self.gripper_goal = [0.0, 0.0]
        else:
            rospy.logerr("Invalid goal")
            return
        
        self.arm_move_group_.go(self.arm_goal, wait=True)
        self.gripper_move_group_.go(self.gripper_goal, wait=True)

        self.arm_move_group_.stop()
        self.gripper_move_group_.stop()

        if self.as_.is_preempt_requested():
            rospy.loginfo("%s is Preempted" % self.action_name_)
            self.as_.set_preempted()
            success = False
        
        if success:
            self.result.success = True
            rospy.loginfo("%s Succeeded" % self.action_name_)
            self.as_.set_succeeded(self.result)


if __name__ == "__main__":
    rospy.init_node("task_server")
    server = TaskServer('task_server')
    rospy.spin()

        

File: src/arduinobot_remote/src/task_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include "arduinobot_remote/ArduinobotTaskAction.h"
#include <moveit/move_group_interface/move_group_interface.h>
#include <vector>

class TaskServer
{
    private:
        ros::NodeHandle nodeHandle;
        actionlib::SimpleActionServer<arduinobot_remote::ArduinobotTaskAction> actionServer;
        std::string actionName;
        arduinobot_remote::ArduinobotTaskResult result;
        std::vector<double> armGoal;
        std::vector<double> gripperGoal;
        moveit::planning_interface::MoveGroupInterface armMoveGroup;
        moveit::planning_interface::MoveGroupInterface gripperMoveGroup;

    public:
        TaskServer(std::string name) :
                  actionServer(nodeHandle, name, boost::bind(&TaskServer::executeCb, this, _1), false),
                  actionName(name),
                  armMoveGroup("arduinobot_arm"),
                  gripperMoveGroup("arduinobot_hand")
        {
            actionServer.start();
        }

        void executeCb(const arduinobot_remote::ArduinobotTaskGoalConstPtr &goal) 
        {
            bool success = true;

            if (goal->task_number == 0)
            {
                armGoal = {0.0, 0.0, 0.0};
                gripperGoal = {-0.7, 0.7};
            }
            else if (goal->task_number == 1)
            {
                armGoal = {-1.14, -0.6, -0.07};
                gripperGoal = {0.0, 0.0};
            }
            else if (goal->task_number ==2)
            {
                armGoal = {-1.57, 0.0, -1.0};
                gripperGoal = {0.0, 0.0};
            }
            else 
            {
                ROS_ERROR("Invalid goal");
                return;
            }

            armMoveGroup.setJointValueTarget(armGoal);
            gripperMoveGroup.setJointValueTarget(gripperGoal);

            armMoveGroup.move();
            gripperMoveGroup.move();

            armMoveGroup.stop();
            gripperMoveGroup.stop();

            if (actionServer.isPreemptRequested() || ros::ok())
            {
                ROS_INFO("%s", actionName.c_str());
                actionServer.setPreempted();
                success = false;
            }

            if (success) 
            {
                result.success = true;
                ROS_INFO("%s Succeeded", actionName.c_str());
                actionServer.setSucceeded();
            }
        }
};


int main(int argc, char** argv)
{
    ros::init(argc, argv, "task_server");
    TaskServer server("task_server");
    ros::spin();
    return 0;
}